modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/components/PageRenderer/index.js

```diff
@@ -0,0 +1,233 @@
+import {Layout} from './Layout.es';
+import * as PageRenderer from './PageRenderer.es';
+import * as Paginated from './Paginated.es';
+import * as SinglePage from './SinglePage.es';
+import * as SuccessPage from './Success.es';
+import * as Tabbed from './Tabbed.es';
+import * as Wizard from './Wizard.es';
+import templates from './index.soy';
+
+const LAYOUT_TYPES = {
+	PAGINATED: 'paginated',
+	SINGLE_PAGE: 'single-page',
+	SUCCESS: 'success',
+	TABBED: 'tabbed',
+	WIZARD: 'wizard',
+};
+
+const LAYOUT_COMPONENTS_TYPES = {
+	[LAYOUT_TYPES.PAGINATED]: Paginated,
+	[LAYOUT_TYPES.SINGLE_PAGE]: SinglePage,
+	[LAYOUT_TYPES.SUCCESS]: SuccessPage,
+	[LAYOUT_TYPES.TABBED]: Tabbed,
+	[LAYOUT_TYPES.WIZARD]: Wizard,
+};
+
+const PAGE_HEADER_TYPES = {
+	EDITABLE: 'editable',
+};
+
+const PAGE_HEADER_COMPONENT_TYPE = {
+	[PAGE_HEADER_TYPES.EDITABLE]: EditablePageHeader.PageHeader,
+};
+
+const isEmptyPage = ({rows}) => {
+	let empty = false;
+
+	if (!rows || !rows.length) {
+		empty = true;
+	}
+	else {
+		empty = !rows.some(({columns}) => {
+			let hasFields = true;
+
+			if (!columns) {
+				hasFields = false;
+			}
+			else {
+				hasFields = columns.some((column) => column.fields.length);
+			}
+
+			return hasFields;
+		});
+	}
+
+	return empty;
+};
+
+const normalizePage = (page, editingLanguageId) => {
+	if (core.isObject(page.description)) {
+		page = {
+			...page,
+			description: page.description[editingLanguageId],
+		};
+	}
+	if (core.isObject(page.title)) {
+		page = {
+			...page,
+			title: page.title[editingLanguageId],
+		};
+	}
+
+	return page;
+};
+
+const getVariant = (page, paginationMode, pages) => {
+	if (
+		page.contentRenderer &&
+		page.contentRenderer !== null &&
+		page.contentRenderer !== ''
+	) {
+		return page.contentRenderer;
+	}
+	else if (paginationMode === null || paginationMode === '') {
+		if (pages.length > 1) {
+			return LAYOUT_TYPES.WIZARD;
+		}
+		else {
+			return LAYOUT_TYPES.SINGLE_PAGE;
+		}
+	}
+	else {
+		return paginationMode;
+	}
+};
+
+/**
+ * Renderer is the main component for dealing with Forms `pages` structure,
+ * you can use the renderer to render fields from different layouts.
+ *
+ * Renderer supports some rendering modes by default
+ * - paginated
+ * - single-page
+ * - tabbed
+ * - wizard
+ *
+ * You can replace all major components or just a few to change the layout
+ * structure, consider the following components
+ * - Container
+ * - Page
+ * - Rows
+ * - Row
+ * - Column
+ *
+ * Use the `overrides` API to pass the components you want to replace,
+ * an example of Page.
+ * <Renderer
+ * 	overrides={{
+ * 	  Page: ({children}) => children
+ * 	}}
+ * />
+ */
+const Renderer = ({
+	activePage = 0,
+	editable = false,
+	editingLanguageId,
+	onBlur,
+	onChange,
+	onFocus,
+	overrides = {},
+	page: pageUnnormalized,
+	pageIndex = 0,
+	pages,
+	paginationMode,
+	readOnly,
+	showSubmitButton,
+	strings,
+	submitLabel,
+}) => {
+	const empty = isEmptyPage(pageUnnormalized);
```

Comment:
I think `unnormalized` term isn't appropriated. Can we proceed with `pageNotNormalized`?

Reply:
Yeah, I was going to change the name but I ended up forgetting. I will try to see other better names.